### PR TITLE
(PUP-7277) Add Init[T] type to the Puppet Type System.

### PR DESCRIPTION
**Pull request by thallgren**
This PR contains the necessary changes to add a new `Init[T]` type to the Puppet Type System. Several minor issues were detected during the implementation work and when adding the unit tests, hence the 9 (maint) commits. Some refactoring was also made on how loaders are handled by types to avoid loader confusion when initializing types using the `Init` type.

### CODE REVIEW

File: lib/puppet/parser/functions/new.rb [hunk 1]
```diff
@@ -942,6 +942,33 @@
...
 * Binary type since 4.8.0
 
+Creating an instance of a `Type` using the `Init` type.
+-------
+
+The type `Init[T]` describes a value that can be used when instantiating a type. When used as the first argument in a call to `new`, it
+will dispatch the call to its contained type and optionally augment the parameter list with additional arguments.
+
+**Example:** Creating an instance of Integer using Init[Integer]
+```
```
Comment (hlindberg) [inline lib/puppet/parser/functions/new.rb]: you want " puppet" after the three backticks to set the syntax

**Comment by thallgren** [comment]
The types `Optional[T]` (which essentially is a `Variant[Undef,T]`) and `Variant[T1, T2]` are not allowed in an Init. A `Variant[Init[T1], Init[T2]]` is OK although it cannot be used as an argument to #new.

**Comment by hlindberg** [comment]
AppVeyor fails with "Debug: Failed to load library 'syslog" seems unrelated.

**Comment by thallgren** [comment]
That's not a failure. It's a debug trace which has been there for a long time.

The curious thing is that although the tests end with "21491 examples, 0 failures, 59 pending", the exit status is still 1. I'll restart and see what happens.

**Pull request merged**